x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportPreviewDataFrameTransformAction.java

```diff
@@ -103,17 +137,40 @@ private void getPreview(Pivot pivot, SourceConfig source, ActionListener<List<Ma
                     pivot.buildSearchRequest(source, null, NUMBER_OF_PREVIEW_BUCKETS),
                     ActionListener.wrap(
                         r -> {
-
                             try {
                                 final CompositeAggregation agg = r.getAggregations().get(COMPOSITE_AGGREGATION_NAME);
                                 DataFrameIndexerTransformStats stats = DataFrameIndexerTransformStats.withDefaultTransformId();
                                 // remove all internal fields
-                                List<Map<String, Object>> results = pivot.extractResults(agg, deducedMappings, stats)
-                                        .peek(record -> {
-                                            record.keySet().removeIf(k -> k.startsWith("_"));
-                                        }).collect(Collectors.toList());
 
-                                listener.onResponse(results);
+                                if (pipeline == null) {
+                                    List<Map<String, Object>> results = pivot.extractResults(agg, deducedMappings, stats)
+                                        .peek(doc -> doc.keySet().removeIf(k -> k.startsWith("_")))
+                                        .collect(Collectors.toList());
+                                    listener.onResponse(results);
+                                } else {
+                                    List<Map<String, Object>> results = pivot.extractResults(agg, deducedMappings, stats)
+                                        .map(doc -> {
+                                            Map<String, Object> src = new HashMap<>();
+                                            String id = (String) doc.get(DataFrameField.DOCUMENT_ID_FIELD);
+                                            doc.keySet().removeIf(k -> k.startsWith("_"));
+                                            src.put("_source", doc);
+                                            src.put("_id", id);
+                                            return src;
+                                        }).collect(Collectors.toList());
+                                    try (XContentBuilder builder = jsonBuilder()) {
+                                        builder.startObject();
+                                        builder.field("docs", results);
+                                        builder.endObject();
+                                        var pipelineRequest = new SimulatePipelineRequest(BytesReference.bytes(builder), XContentType.JSON);
+                                        pipelineRequest.setId(pipeline);
+                                        ClientHelper.executeWithHeadersAsync(threadPool.getThreadContext().getHeaders(),
+                                            ClientHelper.DATA_FRAME_ORIGIN,
+                                            client,
+                                            SimulatePipelineAction.INSTANCE,
```

Comment:
The action name for this is `cluster:admin/ingest/pipeline/simulate`, so previewing a data frame transform with a pipeline now requires that privilege.  We should document this and also test how easy it is to understand the resulting error message if you have all the other privileges required to use the data frame transform _except_ this one.

It's not ideal that using the data frame transform for real does not require any special privilege over and above being able to index data, but simulating does.  The cluster privilege that includes `cluster:admin/ingest/pipeline/simulate` is `manage_ingest_pipelines`, which is the same one that lets you delete ingest pipelines!  As a followup maybe we should enquire about the possibility of adding a `simulate_ingest_pipelines` cluster privilege that just lets you simulate, and not CRUD.

We also need to pass this information on to the UI team when the ability to add a pipeline gets added to the UI, as the UI might want to proactively prevent people who cannot simulate a pipeline from creating a data frame transform that uses one.

Reply:
@droberts195 what do you think of not using the user headers for `_simulate`? I am not sure the reason behind requiring the special permission when they can already index documents referring to the pipeline itself....

Reply:
Yes, that's a good idea.  If we run this action as the internal `_xpack` user then it bypasses the need for the end user to have pipeline admin privileges.  We're still requiring that they have permission to preview the data frame transform.  If we consider simulating the pipeline an internal implementation detail of previewing the data frame transform then it's reasonable to run it as `_xpack`.